### ISSUE TITLE
Official Debian packages for domoticz?

### ISSUE THREAD

**Issue by lnussbaum** [domoticz/domoticz#178]
Hi,

I'm a happy domoticz user, and involved in Debian. I was looking into creating official domoticz packages for Debian. I could maintain backport packages as well.

Is that something you would be interested in?

I haven't looked in detail yet, but it could require some changes to the building infrastructure on the domoticz to make things easier on the packaging side. That's most likely the only thing I'll need from your side.

Please let me know what you think,

Lucas

**Comment by gizmocuz** [comment]
Hello Lucas,

Sounds very good ! We also just using Travis-CI and AppVeyor.

Is this something that can be integrated with travis ?

Let us know the build changes you need to change

**Comment by lnussbaum** [comment]
Hi!

Regarding travis support for building Debian packages, that's likely. I'll look at that later.

I've done some preliminary work to support building domoticz as a Debian package. My work is in
https://github.com/lnussbaum/domoticz

The main obstacle, as far as I can see, is that you currently install everything under /opt. Is there support for installing domoticz following the Linux File Hierarchy Standard, that is, executables in /usr/bin, support data in /usr/share, etc? If not, would you be willing to add such a support?

I've uploaded the result of Debian's lintian tool, that checks for common problems in Debian packages, to http://blop.info/pub/lintian-domoticz.txt. I haven't looked at all the issues listed there, and many of them can probably be fixed on the packaging side, but some of them might be worth fixing on your side as well. Anyway, the most important thing to do is to figure out a strategy to deal with the install paths problem.

Thanks

**Comment by gizmocuz** [comment]
What is wrong with /opt ? everything is installed there, very easy to maintain and update right ?
via command line arguments other paths can be set... but everything in 1 folder, very easy i would say ;)

**Comment by lnussbaum** [comment]
Well installing stuff using the FHS is not negociable for Debian packages. That's part of the role of Linux distributions to standardize on such things over all the software they ship.

**Comment by gizmocuz** [comment]
Okey, i think this is already possible by using the command line arguments:

-wwwroot file_path (for example /opt/domoticz/www)
-dbase file_path (for example /opt/domoticz/domoticz.db)
-userdata file_path (for example /opt/domoticz)

**Comment by lnussbaum** [comment]
Hi,

yes, this could work (installing in a custom location + changing the default search path in domoticz).
I'll look into that in a few weeks: I broke my arm in the meantime.

**Comment by gizmocuz** [comment]
Ai... good luck with your arm... painful when typing

**Comment by empierre** [comment]
@lnussbaum the domoticz image is on a raspbian (no support of armv6 in debian) thus all is compiled for armv6, and it works well on armv7 (PI2, Odroid...). Packages compiled for armv7 segfault on a PI A/B/+ (-thus armv6)

**Comment by axello** [comment]
No news on this? 
I would love to be able to “apt-get upgrade” instead of compiling from source on my Rasp-Pi.
Or setting up a cross-compile and remote deployment.

**Comment by dwmw2** [comment]
There's a small amount more to it than the existing command line options can manage. OpenWRT has a patch at https://github.com/openwrt/packages/blob/master/utils/domoticz/patches/903_fhs.patch which I need to clean up and try submitting here.

**Comment by dwmw2** [comment]
I've *started* cleaning that one up at https://github.com/dwmw2/packages/blob/domoticz/utils/domoticz/patches/903_fhs.patch but wanted to split it up a bit into separate components and review it myself, before making a PR here. You might as well look though...

**Comment by dwmw2** [comment]
ping @stintel

**Comment by FedericoCeratto** [comment]
@gizmocuz we are interested in packaging domoticz in Debian and FreedomBox. Would you consider reopening the issue?

**Comment by gizmocuz** [comment]
@FedericoCeratto, there is no bug, this is a feature request/issue
But at the present you can configure everything, all paths are reconfigurable at startup
what might be a good idea is, instead of command line options, support /etc/domoticz.conf ?
The install script could create this configuration file in real-time ?

**Comment by FedericoCeratto** [comment]
The packaging files will be https://salsa.debian.org/debian/domoticz/tree/master/debian

Paths and other configuration are supplied when invoking the executable: https://salsa.debian.org/debian/domoticz/blob/master/debian/domoticz.service

The HTTP and HTTPS ports could be made configurable using a /etc/default/domoticz file assuming domoticz cannot parse /etc/domoticz.conf by itself.

(Related ITP  https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=899058 )

**Comment by gizmocuz** [comment]
@FedericoCeratto, there are more options like specifying the HTTPS certificate paths.
The command line option in domoticz.service might get to big and unreadable on small screens
This is the reason why in the current startup script we can add them on multiple lines
But i am totally for a 'service' (specially for a restart option) and a configuration file
the basic configuration file (/etc/domoticz.conf) may already include all possible options with commands but commented out.

**Comment by dwmw2** [comment]
It's OK to require command-line options. If we want to look at changing the defaults for them to reduce command line size, let's do that in a later stage. It would be nice to start by just tidying up the last details to make sure we *can* do the right thing.

Firstly, we really want to link against *system* libraries not statically link our own. I actually fixed a bunch of those a year or two ago when I added `USE_BUILTIN_LUA`, `USE_BUILTIN_SQLITE` and `USE_STATIC_LIBSTDCXX` build options. Are there more we want to do now? 

Then there's the [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) support. This is really important for OpenWRT where only part of the file system is actually writeable at all. We probably want to use /var/lib/domoticz for `szUserDataFolder` (-userdata) and /usr/share/domoticz for `szStartupFolder` (-approot). Most of that works already; I think we just want to fix up the dzVents directories so that `dzVents/runtime` is found under `szStartupFolder` not `szUserDataFolder`.

We also probably need to fix the fact that floorplans live in the "read-only" wwwroot directory too. Those need to live under `szUserDataFolder`.

@stintel you also added a `-scripts` command line option but I think we can drop that. As long as the dzVents runtime is in the right place, scripts under szUserDataFolder is fine.

**Comment by gizmocuz** [comment]
@dwmw2, are we 100% sure that on default debian/linux installs those system libraries exists ?
Else after an upgrade all current Domoticz instances will stop working.
Next, the newer Lua does not seem to work with Domoticz, using a system library will/might cause it to stop.
It is also a tiny addon to the exe, so what is the main problem ? (For windows of course we need to keep it)
Libstd, yeah... so the build system is compiling with GCC 17.1 , the users system has GCC 4.2, that will work ?

FHS, that sounds more then right! Can you make a PR for this ?

Floorplan, ehhh as discussed 3 years ago (i think) those images 'Should' be stored in the database. (like the switch icons)
This for obvious reasons (backup/restore)
Hopefully someone will create a PR that will do a database migration and moves them in the database, i could do this if you want ?
Yes the database could get larger depending on the image size, but this we can limit in the floorplan editor, force png or jpg)

**Comment by dwmw2** [comment]
The linking against system libraries isn't really for *your* builds, so it shouldn't matter if the libraries are installed by default. I assume you'll continue to make self-contained static builds, which will be the default. This is for *package* builds, for Fedora .rpm packages, Debian .deb packages and OpenWRT .ipkg packages. Those have dependencies so that if you install an application package it also installs the libraries that application is linked against.

Dynamic linking is *particularly* important for OpenWRT which is often space-constrained _(but my router makes an *excellent* little box for running Domoticz on)_. And for Fedora where packaging rules mandate that applications SHOULD NOT bundle their own copies of libraries (partly for security reasons, so there's *one* copy of the library to be updated with fixes, not dozens all statically linked into various packages).

But also, the linking is mostly a solved problem, a year or three ago in #203. I was just asking if there are *more* libraries we need to be able to link dynamically, since I last looked. Lua can be problematic, as you say — I think OpenWRT *is* actually statically linking liblua for that reason.

For the FHS thing... I *think* it's just the dzVents thing as I said. I need to do some more testing on the top commit (currently d09d20cca5952) in https://github.com/dwmw2/domoticz/commits/openwrt — and I haven't even looked hard at the three commits behind it yet, which are from @stintel

If you want to do the floorplan→database migration thing that would be much appreciated; thanks. It's code I'm not stunningly familiar with so you'd be a lot quicker than me. Thanks.

**Comment by gizmocuz** [comment]
Okey that makes sense , but what will happen if for instance OpenWRT starts to ship with lua 18, and domoticz needs lua 5 ? for sure the system will not replace lua 18 with 5 ?
(Just trying to avoid dll nightmares)

I will try to make a PR for the floorplan and let you know for testing asap

**Comment by dwmw2** [comment]
That already happened, IIRC — although the numbers were different. OpenWRT used a static liblua for the Domoticz build. Then the version of lua in OpenWRT was updated, and Domoticz didn't work with the new one, so we switched the Domoticz package to use a built-in lua instead.

DLL nightmares don't happen on sane systems because we have package dependencies to prevent them :)

**Comment by gizmocuz** [comment]
i am not sure that is 100% true.
mysuperdll.so is actually a sumlink mysuperdll.4.3.2.1.so
So if the system is depending on version 4321, and domoticz on 7890, then there might rise an issue here
Or am i missing something here ?

**Comment by dwmw2** [comment]
Yeah..  you're missing sonames. On phone right now so a little brief but basically...

mysuperdll.so is never used at runtime. It is only ever used for *building*. You actually have *two* symlinks. It looks more like...

mysuperdll.so -> mysuperdll.so.1
mysuperdll.so.1 -> mysuperdll.so.1.4321
mysuperdll.so.1.4321 is the actual library file.

The application knows that it is linked against mysuperdll.so.1.

The *packaging* knows that it requires mysuperdll.so.1.

Now, the library is updated to mysuperdll v7890. There are two options here -- either the change is backward-compatible and existing apps will continue to work, or it isn't.

The common case is that it's just new functionality being added and it's still compatible. So the new library just drops in and you have

libmysuperdll.so.1 -> libmysuperdll.so.1.7890
libmysuperdll.so.1.7890

Applications linked against mysuperdll.so.1 continue to work.

The alternative is if an *incompatible* change has been made -- arguments to an existing function are changed, or some app-visible structure is changed in a way that breaks binary compatibility. In that case, the library author changes the soname to show the incompatibility. You get...

libmysuperdll.so.2 -> libmysuperdll.so.2.7890
libmysuperdll.so.2.7890

Note that both versions can coexist, and different apps can run against both at the same time. Only the bare .so symlinks used for *compiling* (well, linking) are mutually exclusive.

The packaging also knows, so if you try to install an app which requires mysuperdll.so.1 and your system has moved on to mysuperdll.so.2, you get a dependency error (or the old library installed in parallel).

**Comment by gizmocuz** [comment]
Thanks for the explanation! Learned something again ;)
So it might go wrong when the system upgraded from version 1 to 2... but then there might be a new domoticz package ;)

**Comment by dwmw2** [comment]
Right. Part of the process of updating a library with an incompatible change and a new soname, is to make sure all the dependent packages are updated too (and potentially provide a compatibility version of the lib if that's too hard, like with OpenSSL 1.0->1.1)

**Comment by dwmw2** [comment]
Back to a real keyboard... so there's one part I didn't completely explain. You asked what happens if Domoticz is built against v7890 and then installed on a system with v1234.

If v7890 is incompatible and has a new soname then that's fine; the Domoticz package will require the new library. But... what if v7890 *didn't* change the soname? What if it only added one new function, for example, so it was still backward compatible? But what if Domoticz *uses* that new function from libmysuperdll.so.1.7890, which isn't present in the libmysuperdll.so.1.1234 that's on the running system?

Mostly, the answer has been "don't do that then". The soname thing alone doesn't help here. And really, it just *doesn't* happen in practice. Distribution packages are built against the version of the library which is in that distro at the time, and we don't *downgrade* the libraries in the distro. The soname thing helps you cope with *upgrading* libraries, not with downgrading them. When building an application, you need to build it against the *oldest* version of the library that it might end up being run against.

But there *is* a technical solution to that too, which is symbol versioning. A well-maintained library will actually mark symbols with the version they first appeared in — a minor version, the "1.7890" version not just the major version which is encoded in the soname.

So, here's an example library I happen to be intimately familiar with...
```
nm -D --defined-only --with-symbol-versions /usr/lib64/libopenconnect.so.5.4.0
…
000000000002d860 T openconnect_get_peer_cert_chain@@OPENCONNECT_5_3
000000000002d020 T openconnect_get_peer_cert_DER@@OPENCONNECT_5.0
```
The `openconnect_get_peer_cert_chain` function was added in the v5.3 ABI, but since it was only an additional function, the soname *wasn't* changed from `libopenconnect.so.5` — because existing applications would still work with the new library. So we have the situation described above: an application *might* be built against the new library, might *use* the new function, and sonames alone wouldn't help it. So what does? 

This is where the symbol versions come in — the special `@@OPENCONNECT_5_3` metadata at the end of the symbol name. It's all invisible; app developers never really notice it and just use the standard function names. But the linker and the packaging software knows, so the dependency can be encoded correctly into the RPM package for example:
```
$ rpm -q --requires openconnect | grep OPENCONN
libopenconnect.so.5(OPENCONNECT_5.0)(64bit)
libopenconnect.so.5(OPENCONNECT_5_1)(64bit)
libopenconnect.so.5(OPENCONNECT_5_2)(64bit)
libopenconnect.so.5(OPENCONNECT_5_3)(64bit)
libopenconnect.so.5(OPENCONNECT_5_4)(64bit)
libopenconnect.so.5(OPENCONNECT_PRIVATE)(64bit)
```

So if I try to install this binary package against an older version of the library which *doesn't* provide `libopenconnect.so.5(OPENCONNECT_5_3)` then RPM will refuse because the dependency isn't met.

Here endeth the second lesson ... :)

**Comment by gizmocuz** [comment]
Pfff this took some time/testing, but i think i now migrated the floorplan stuff to database only.
If it possible to test this ?

https://github.com/domoticz/domoticz/pull/2400

**Comment by dwmw2** [comment]
The dzVents runtime directory fix is #2401

**Comment by dwmw2** [comment]
Submitted a handful of other things too from the OpenWRT packaging.
https://github.com/openwrt/packages/pull/6091

**Comment by FedericoCeratto** [comment]
@gizmocuz : is there any simple way to disable the upgrade check at build time or programmatically? Also, any way to pre-configure a couple of example items in the dashboard? Thanks

**Comment by dwmw2** [comment]
OpenWRT used to have a patch to do the former; it seems to have got lost. Looked something like this:
```
--- a/main/SQLHelper.cpp
+++ b/main/SQLHelper.cpp
@@ -2688,7 +2688,7 @@ bool CSQLHelper::OpenDatabase()
 
        if (!GetPreferencesVar("UseAutoUpdate", nValue))
        {
-               UpdatePreferencesVar("UseAutoUpdate", 1);
+               UpdatePreferencesVar("UseAutoUpdate", 0);
        }
 
        if (!GetPreferencesVar("UseAutoBackup", nValue))
```

The latter you'd probably have to do by prepopulating virtual hardware devices in `domoticz.db` (in which case you could prepopulate the `UseAutoUpdate` pref too and not bother with the above hack).

**Comment by gizmocuz** [comment]
@FedericoCeratto, we could make a switch for it... let me see what i can do, because you also do not want to see it in the menu/settings i suppose ;)

**Comment by gizmocuz** [comment]
@FedericoCeratto , new commit, it is now possible to use a new command line option "-noupdates"
It will also hide the two options (Update/Check for Update) from the Menu Setup dropdown, and the update settings from the setup page

**Comment by dwmw2** [comment]
@gizmocuz that's great; thanks. I'll roll that into the OpenWRT builds too.

**Comment by FedericoCeratto** [comment]
@gizmocuz it would be good to optionally disable the static, built-in zip library as well.

**Comment by gizmocuz** [comment]
@FedericoCeratto , which library you are referring about ?

**Comment by dwmw2** [comment]
The one in `zip/` ... which is weird because we dont build `unzip.cpp` and we seem to have appended its contents to `unzip.h` and made everything inline. Which confused me for a while...

If we fix that to actually build `unzip.cpp` normally instead of inlining it all, we could use an external libminizip.

**Comment by gizmocuz** [comment]
zip is not used in this project. But it is possible to compress the whole www folder in a encrypted zip file

**Comment by dwmw2** [comment]
We use it for uploading custom icons.

**Comment by gizmocuz** [comment]
Thats correct true... i dont suspect this to be any of significant size ?

**Comment by dwmw2** [comment]
> Thats correct true... i dont suspect this to be any of significant size ?

As it happens, I can answer that one... it's about 20KiB on x86_64.

```
$ size domoticz
   text	   data	    bss	    dec	    hex	filename
21116426	  53768	  50988	21221182	143cf3e	domoticz
[dwoodhou@i7 domoticz]$ cmake -DUSE_BUILTIN_MINIZIP=NO
...
$ size domoticz
   text	   data	    bss	    dec	    hex	filename
21097058	  53872	  50988	21201918	14383fe	domoticz
```

But linking to the existing shared libraries isn't just a matter of size. It's about ensuring that security updates get deployed correctly. If a buffer overflow with a malicious `.zip` file is found in minizip, a new minizip library is shipped. Are they going to update the Domoticz package too? History says that static libraries built into other applications are *often* not updated when they should be, and even if we were to manage to track them it's still a lot of extra work and extra packages to be updated. So some distributions have packaging guidelines which say that you SHOULD NOT include static versions of any libraries which were already packaged for that distribution.

**Comment by dwmw2** [comment]
Given the ABI breakage at https://github.com/eclipse/mosquitto/issues/829 we should probably look at ditching libmosquittopp too. It's fairly much the *most* gratuitous use of C++ I've ever had the misfortune to witness, anyway. It's a trivial and pointless wrapper around the C ABI, and we'd be better off using that directly. The *only* thing the C++ wrapper really does, is break the ABI for us and cause crashes which wouldn't have happened if we were using libmosquitto directly.

**Comment by FedericoCeratto** [comment]
FYI the main reason for distributions to deduplicate/unvendorize dependencies is security as @dwmw2 pointed out. Another reason is to simplify and automate handling copyright attribution (and avoid infringement). Also a packaged library used by many applications gets a lot of testing, improving reliability.

**Comment by gizmocuz** [comment]
I do not see where we could have an overflow in this library.
It is a small footprint, and the last update was from 22 September 2004
It is also depended on gzip
And we use only the headers...
Just leave it be...

**Comment by dwmw2** [comment]
We don't really "use only the headers". Strictly speaking I suppose that's true... but only because we *copied* all the C code into the `.h` file and marked all the functions `inline`. I don't think that really counts :)

**Comment by dwmw2** [comment]
And there have actually been buffer overflows in the `contrib/minizip` code fixed in upstream zlib, since 2004: https://github.com/madler/zlib/commit/601b542a9d6c3689fce1414ec089d5f193656faa

Admittedly that's not in a part of the code that we're using, but the general policy remains: distributions *really* want us to be using dynamic libraries — not because of the historical bugs, but because of *future* ones.

**Comment by gizmocuz** [comment]
No it is not in the part of the code we are using

**Comment by gizmocuz** [comment]
Merged as well

**Comment by dwmw2** [comment]
Thanks.

**Comment by dpocock** [comment]
here are the latest links about the status of packages

blog / instructions:
https://danielpocock.com/domoticz-packaging-debian-ubuntu-raspbian-fedora/

discussion:
http://www.domoticz.com/forum/viewtopic.php?f=47&t=28951

**Comment by gizmocuz** [comment]
@dpocock , not sure what you a referring too, this is a issue from 2018! ... what seems to be the issue ?
